### PR TITLE
Dicom adapter routing extension to retry

### PR DESCRIPTION
**Pull request by NikolaiKovalenko**
CStoreService logic reverted to 'before retry branch';
CStoreService defining of defaultDicomWebClient and destination filtering logic moved into separate class IDestinationClientFactory and its impls;
Added 3 configuration mode for CStoreService:
1. without backup, without multiple routing;
2. with backup, with single mode (standard destination filtering - get first occurrence);
3. with backup, with multiple mode (destination filtering with multiple routing ability).
Changed container for the filterToDestination changed from Map<filter, destination> to List<Pair<filter, destination>> for new multiple dest upload logic impl.
unit and integration tests added;

### CODE REVIEW

File: import/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/ImportAdapter.java
Comment (danielbeaudreau) [inline import/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/ImportAdapter.java:132]: Let's throw an error if sendToAllMatchingDestinations is true but the backup service is null. Since we require the backup service to send to multiple destinations

File: import/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/ImportAdapter.java
Comment (danielbeaudreau) [inline import/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/ImportAdapter.java:157]: multipleDestinationSendService should only be passed into the constructor if the user is using multiple destinations. Otherwise this should be null, because the multiple destination send service will not work without a backup

File: import/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/cstore/multipledest/MultipleDestinationUploadService.java
Comment (danielbeaudreau) [inline import/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/cstore/multipledest/MultipleDestinationUploadService.java:40]: Could we initialize this only if len(dicomDestinations) > 0 to save on resources / connections?

File: import/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/cstore/backup/BackupState.java
Comment (danielbeaudreau) [inline import/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/cstore/backup/BackupState.java:5]: I'm wondering why does this have to be volatile? Maybe we should copy BackupState instead of sharing one object to eliminate the need for volatile?

File: import/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/cstore/backup/BackupUploadService.java
Comment (danielbeaudreau) [inline import/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/cstore/backup/BackupUploadService.java:62]: I suggest renaming sopClassUidBackupState to just sopClassUid to make it more clear what this is

File: import/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/cstore/backup/BackupUploadService.java
Comment (danielbeaudreau) [inline import/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/cstore/backup/BackupUploadService.java:163]: I think we could miss logging this error if the try block throws an exception above.
I think we should move this to the beginning of the outer catch block after logUploadFailed(io)

File: import/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/cstore/backup/BackupUploadService.java
Comment (danielbeaudreau) [inline import/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/cstore/backup/BackupUploadService.java:205]: same here

**Review by danielbeaudreau — CHANGES_REQUESTED** [review]
Thanks. I only had time to review the code changes, not the tests. Will take a look at those next.

File: import/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/cstore/multipledest/MultipleDestinationUploadService.java
Comment (oleksii-opanasenko) [reply import/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/cstore/multipledest/MultipleDestinationUploadService.java:40]: it`s only Factory class instance creation. Connection will be created on create() method call. Check len(dicomDestinations) > 0 adde to MultipleDestinationUploadService. Thank you.

File: import/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/cstore/backup/BackupState.java
Comment (oleksii-opanasenko) [reply import/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/cstore/backup/BackupState.java:5]: it`s a part of fail logic that i have removed before. Fixed.

File: import/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/cstore/backup/BackupUploadService.java
Comment (oleksii-opanasenko) [reply import/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/cstore/backup/BackupUploadService.java:163]: Yes, agree. Have checked and refactored adding of  events by schema: one destination send - one event (add event only on last send attempt)

File: import/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/CStoreService.java
Comment (danielbeaudreau) [inline import/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/CStoreService.java:130]: We need to still support the backup service in the case where we do not send to multiple destinations

**Review by danielbeaudreau — CHANGES_REQUESTED** [review]

File: import/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/cstore/backup/GcpBackupUploader.java
Comment (oleksii-opanasenko) [reply import/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/cstore/backup/GcpBackupUploader.java:131]: will remove, sorry, it`s for tests.

**Review by danielbeaudreau — CHANGES_REQUESTED** [review]
Thanks for the extensive unit test coverage. Will you be adding integration tests after?

File: import/src/test/java/com/google/cloud/healthcare/imaging/dicomadapter/CStoreServiceTest.java
Comment (danielbeaudreau) [reply import/src/test/java/com/google/cloud/healthcare/imaging/dicomadapter/CStoreServiceTest.java:544]: What I meant it, I think this line should be uncommented

File: import/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/ImportAdapter.java
Comment (danielbeaudreau) [reply import/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/ImportAdapter.java:310]: This is our existing behavior for single destination

File: import/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/ImportAdapter.java
Comment (oleksii-opanasenko) [reply import/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/ImportAdapter.java:118]: done. Why we need --dimse_cmove_aet at all ? It used in like an alternative of --dimse_aet for cMove (and cStore now). 
(line 145)

File: import/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/ImportAdapter.java
Comment (oleksii-opanasenko) [reply import/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/ImportAdapter.java:135]: Maybe multipleDestinationSendService  is a bad name, but this service take a part in single mode processing (send_to_all_matching_destinations not set, but backup present ). In this mode SingleDestinationClientFactory delivers to multipleDestinationSendService  only one healthDestination from destMap after filtering. dicomDest list is empty. --> multipleDestinationSendService  processes healthDestination list with one dest.  (it already works in CStore tests, and now i have fixed it in ImportAdapter.)

File: import/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/ImportAdapter.java
Comment (oleksii-opanasenko) [reply import/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/ImportAdapter.java:209]: moved to configureDestinationClientFactory.

File: import/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/CStoreService.java
Comment (oleksii-opanasenko) [reply import/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/CStoreService.java:130]: it`s already present in MultipleDestinationUploadService with SingleDestinationClientFactory. (configuration of this mode fixed in ImportAdapter now)

File: import/src/test/java/com/google/cloud/healthcare/imaging/dicomadapter/CStoreServiceTest.java
Comment (oleksii-opanasenko) [reply import/src/test/java/com/google/cloud/healthcare/imaging/dicomadapter/CStoreServiceTest.java:464]: uncommented. thank you.

File: import/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/ImportAdapter.java
Comment (oleksii-opanasenko) [reply import/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/ImportAdapter.java:310]: could you show me  this behavior please. As i understand, for single dest on filtering we will have JSONException on no string value for the key. https://github.com/GoogleCloudPlatform/healthcare-dicom-dicomweb-adapter/blob/master/util/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/DestinationsConfig.java#L46

File: import/src/test/java/com/google/cloud/healthcare/imaging/dicomadapter/cstore/multipledest/sender/CStoreSenderTest.java
Comment (oleksii-opanasenko) [reply import/src/test/java/com/google/cloud/healthcare/imaging/dicomadapter/cstore/multipledest/sender/CStoreSenderTest.java:74]: it`s not my code. Present in CMove tests too. Do y want to remove it?

File: import/src/test/java/com/google/cloud/healthcare/imaging/dicomadapter/cstore/multipledest/sender/CStoreSenderTest.java
Comment (oleksii-opanasenko) [reply import/src/test/java/com/google/cloud/healthcare/imaging/dicomadapter/cstore/multipledest/sender/CStoreSenderTest.java:145]: actualReceivedBytesByServer contains the header, that i have not.  That  way i equal sub array.

File: import/src/test/java/com/google/cloud/healthcare/imaging/dicomadapter/cstore/multipledest/sender/CStoreSenderTest.java
Comment (danielbeaudreau) [reply import/src/test/java/com/google/cloud/healthcare/imaging/dicomadapter/cstore/multipledest/sender/CStoreSenderTest.java:74]: Hmm, I guess there's no harm in keeping this.

File: import/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/ImportAdapter.java
Comment (danielbeaudreau) [reply import/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/ImportAdapter.java:310]: I think you are right. I guess no one tested that. I'll have to update our docs.

File: import/src/test/java/com/google/cloud/healthcare/imaging/dicomadapter/cstore/multipledest/sender/CStoreSenderTest.java
Comment (oleksii-opanasenko) [reply import/src/test/java/com/google/cloud/healthcare/imaging/dicomadapter/cstore/multipledest/sender/CStoreSenderTest.java:110]: fixed.

**Comment by oleksii-opanasenko** [comment]
@danielbeaudreau 
1. all fixes done.
2. I can write tests for ImportAdapter class to fix all types of cStore configurations  (multiple / backup / without) if you want. But i need to move all configure methods to some factory class / do they visible for testing.
3. Nikolai will do the commit with integration test in near days. (We have a new story to update integration test approach, do we need write oldStyle tests now ? )

**Comment by danielbeaudreau** [comment]
Thanks! I will review this soon (tomorrow morning). 
For #2, I think this is a good idea, but let's defer to a follow up PR after this one is merged into the branch since this PR already contains a lot of changes.
For #3, you do not need to write them in the old style, feel free to wait until that story is complete before writing them.

File: import/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/ImportAdapter.java
Comment (danielbeaudreau) [reply import/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/ImportAdapter.java]: It was added in case the user wants a separate AE title when sending rather than when receiving.
Whether or not its useful to users, I'm not sure.

File: import/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/cstore/backup/GcpBackupUploader.java
Comment (danielbeaudreau) [inline import/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/cstore/backup/GcpBackupUploader.java:127]: The getCredentialOldStyle method isn't used anymore, please delete it

File: import/src/test/java/com/google/cloud/healthcare/imaging/dicomadapter/cstore/multipledest/sender/CStoreSenderTest.java
Comment (danielbeaudreau) [inline import/src/test/java/com/google/cloud/healthcare/imaging/dicomadapter/cstore/multipledest/sender/CStoreSenderTest.java:127]: I'm trying to understand why we do a cmove here?

**Review by danielbeaudreau — COMMENTED** [review]
Thanks

**Comment by danielbeaudreau** [comment]
A test is currently failing:

com.google.cloud.healthcare.imaging.dicomadapter.cstore.multipledest.sender.CStoreSenderTest > sendFile FAILED
    java.nio.file.NoSuchFileException at CStoreSenderTest.java:127
    java.lang.NullPointerException at CStoreSenderTest.java:159

Are you able to see the test logs below?

File: import/src/test/java/com/google/cloud/healthcare/imaging/dicomadapter/cstore/multipledest/sender/CStoreSenderTest.java
Comment (oleksii-opanasenko) [reply import/src/test/java/com/google/cloud/healthcare/imaging/dicomadapter/cstore/multipledest/sender/CStoreSenderTest.java]: I needed any dicom test endpoint with stub. This one i found in tests that already implemented.

Comment (danielbeaudreau) [reply import/src/test/java/com/google/cloud/healthcare/imaging/dicomadapter/cstore/multipledest/sender/CStoreSenderTest.java]: Could we remove the code that calls cmove on the association and just call cstore in the sendFile test? Isn't the stub for cstore all we need?

Comment (oleksii-opanasenko) [reply import/src/test/java/com/google/cloud/healthcare/imaging/dicomadapter/cstore/multipledest/sender/CStoreSenderTest.java]: yes, fixed.

**Review by danielbeaudreau — CHANGES_REQUESTED** [review]
Thanks! two very minor comments, and then this is good to merge.

Question around testing, for scenario #3 mentioned in the original pull request comment, have you been able to test this manually under the following conditions? Since we don't have an integration test for this yet, I just wanted to make sure it is tested before merging.

- multiple healthcare destinations
- multiple dicom destinations
- using gcs backup
- using local backup

File: import/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/cstore/backup/DelayCalculator.java
Comment (danielbeaudreau) [inline import/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/cstore/backup/DelayCalculator.java:26]: This is a bit of a nit pick, but for attemptsLeft < 1, I think we should return the maxWaitingTimeBtwUpload, and for attemptsLeft > attemptsAmount we should return the minimum.

I know neither scenario should ever happen, but I think these defaults should be swapped.

**Review by danielbeaudreau — APPROVED** [review]
Thanks, just waiting on testing conformation to merge this.

**Comment by oleksii-opanasenko** [comment]
please wait. doing another one manual test for double check

**Comment by oleksii-opanasenko** [comment]
manual scenarios test success:
--send_to_all_matching_destinations + local_backup  --> 2 webDest, 2 dicomDest - OK
--send_to_all_matching_destinations + gcs_backup  --> 2 webDest, 2 dicomDest - OK
multiple dest mapping present + local_backup  --> IllegalArgEx with err msg - OK
--send_to_all_matching_destinations + one of mandatory backup flags absent  --> IllegalArgEx with err msg - OK
sys out logs attached.

**Comment by danielbeaudreau** [comment]
Thanks. 
For "multiple dest mapping present + local_backup --> IllegalArgEx with err msg - OK", which error message does this correspond to? Not sure I understood which test case this is

**Comment by oleksii-opanasenko** [comment]
explanation:  if we remove --send_to_all_matching_destinations  mandatory flag from start app args, but leave multiple DEST filter mapping like: filterA ->dest, filterA -> dest2 we will have Ex:  https://github.com/NikolaiKovalenko/healthcare-dicom-dicomweb-adapter/blob/DICOM_Adapter_Routing_Extension_to_retry/util/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/DestinationsConfig.java#L48 because this method expects single dest filter mapping mode data.

**Comment by danielbeaudreau** [comment]
Got it, thanks. Merging this to master now

**Pull request merged**
